### PR TITLE
Kdn adding and diagnostic filtering

### PR DESCRIPTION
**Pull request by RayMSMS**
1. Add Kdn sugar type in the dictionary
2. Add a new function "DiagnosticIon filter", which uses the diagnostic ion to distinguish the isobaric peak.
    If we find 274/292 peak in Scan, we remove the glycan without A(NeuAc) in the candidate list.
3. Add the corresponding test for the filter function. Isobaric case: H1N1A1 & N2K1.

### CODE REVIEW

**Review by nbollis — CHANGES_REQUESTED** [review]
Looks good overall. I had a few questions and concerns that I left as comments throughout the PR. I also checked the code coverage and did not find an area where it would be easy to test the additions that you made.

File: MetaMorpheus/EngineLayer/Glycan_Mods/OGlycan/Bullfrog 64 O-glycan wKdn corrected_convert.txt
[the remark this answers is not in the document]
Comment (RayMSMS) [reply MetaMorpheus/EngineLayer/Glycan_Mods/OGlycan/Bullfrog 64 O-glycan wKdn corrected_convert.txt:1]: I just added to the existing glyco files. The Glycan_Mods is a default glycan database collection for the user.

File: MetaMorpheus/EngineLayer/GlycoSearch/GlycanDatabase.cs [hunk 1]
```diff
@@ -207,7 +240,7 @@ public static List<GlycanIon> NGlycanCompositionFragments(byte[] kind)
 
                             for (int add_fuc_count = 2; add_fuc_count <= fuc_count; add_fuc_count++)
                             {
-                                GlycanIon add_fuc_glycanIon = ExtendGlycanIon(hexose_glycanIon, 0, 0, (byte)add_fuc_count, 0, glycan_mass);
+                                GlycanIon add_fuc_glycanIon = ExtendGlycanIon(hexose_glycanIon, 0, 0, 1, 0, glycan_mass);
```
[the remark this answers is not in the document]
Comment (RayMSMS) [reply MetaMorpheus/EngineLayer/GlycoSearch/GlycanDatabase.cs:243]: The fragment is for N-glycan fragmentation, we don't use that in O-pair search right now. But, I still do a small revision to correct the fragment issue. A tester is also built.

File: MetaMorpheus/EngineLayer/GlycoSearch/GlycoSearchEngine.cs [hunk 2]
```diff
@@ -15,13 +15,13 @@ namespace EngineLayer.GlycoSearch
     public class GlycoSearchEngine : ModernSearchEngine
     {
         public static readonly double ToleranceForMassDifferentiation = 1e-9;
-        private readonly int OxoniumIon204Index = 9; //Check Glycan.AllOxoniumIons
-        protected readonly List<GlycoSpectralMatch>[] GlobalCsms;
+        private readonly int OxoniumIon204Index = 9;               // Check Glycan.AllOxoniumIons
+        protected readonly List<GlycoSpectralMatch>[] GlobalCsms;  // Why don't we call it GlobalGsms?
```
[the remark this answers is not in the document]
Comment (RayMSMS) [reply MetaMorpheus/EngineLayer/GlycoSearch/GlycoSearchEngine.cs:19]: Got it. I already changed it.

File: MetaMorpheus/EngineLayer/GlycoSearch/GlycoSpectralMatch.cs [hunk 3]
```diff
@@ -357,38 +378,45 @@ public string GlycoToString()
...
+        /// <param name="OGlycanBoxLocalization"> all case of the pair </param>
+        /// <param name="localizationLevel"> level 1 to level 3 </param>
+        /// <returns> A tuple, represent the pair and its confidience ex. [3,5,ture] means glycan 5 located on glycosite 3, and very confidience </returns>
         public static List<Tuple<int, int, bool>> GetLocalizedGlycan(List<Route> OGlycanBoxLocalization, out LocalizationLevel localizationLevel)
         {
             List<Tuple<int, int, bool>> localizedGlycan = new List<Tuple<int, int, bool>>();
 
-            //Dictionary<string, int>: modsite-id, count
-            Dictionary<string, int> seenModSite = new Dictionary<string, int>();
+            Dictionary<string, int> ModSiteSeenCount = new Dictionary<string, int>(); // all possible glycan-sites pair, Dictionary<string, int>: site-glycan pair, count
```
[the remark this answers is not in the document]
Comment (RayMSMS) [reply MetaMorpheus/EngineLayer/GlycoSearch/GlycoSpectralMatch.cs:394]: Got it. Corrected it.

File: MetaMorpheus/EngineLayer/GlycoSearch/GlycanBox.cs [hunk 4]
```diff
@@ -102,14 +120,19 @@ public static IEnumerable<GlycanBox> BuildChildOGlycanBoxes(int maxNum, int[] gl
             }
         }
 
+        /// <summary>
+        /// Constructor of GlycanBox.
+        /// </summary>
+        /// <param name="ids"> The glycanBox composition, each number represent one glycan index in the database</param>
+        /// <param name="targetDecoy"></param>
         public GlycanBox(int[] ids, bool targetDecoy = true):base(ids)
```
[the remark this answers is not in the document]
Comment (RayMSMS) [reply MetaMorpheus/EngineLayer/GlycoSearch/GlycanBox.cs]: Corrected it. Thanks

**Review by trishorts — APPROVED** [review]

**Review by Alexander-Sol — APPROVED** [review]

**Review by nbollis — APPROVED** [review]

**Pull request merged**